packages/flutter/lib/src/material/mergeable_material.dart

```diff
@@ -0,0 +1,562 @@
+  }) : super(key: key);
+
+  /// The children of the [MergeableMaterial].
+  final List<MergeableMaterialItem> children;
+
+  /// The main layout axis.
+  final Axis mainAxis;
+
+  /// The elevation of all the [Material] slices.
+  final int elevation;
+
+  @override
+  String toString() {
+    return 'MergeableMaterial('
+      'key: $key, mainAxis: $mainAxis, '
+      'elevation: $elevation, children: $children'
+    ')';
+  }
+
+  @override
+  _MergeableMaterialState createState() {
+    return new _MergeableMaterialState();
+  }
+}
+
+class _AnimationTuple {
+  _AnimationTuple({
+    this.controller,
+    this.startAnimation,
+    this.endAnimation,
+    this.gapAnimation,
+    this.gapStart: 0.0
+  });
+
+  final AnimationController controller;
+  final CurvedAnimation startAnimation;
+  final CurvedAnimation endAnimation;
+  final CurvedAnimation gapAnimation;
+  double gapStart;
+}
+
+class _MergeableMaterialState extends State<MergeableMaterial> {
+  List<MergeableMaterialItem> _children;
+  Map<LocalKey, _AnimationTuple> _animationTuples;
+
+  void _initGap(MaterialGap gap) {
+    final AnimationController controller = new AnimationController(
+      duration: kThemeAnimationDuration
+    );
+
+    final CurvedAnimation startAnimation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+    final CurvedAnimation endAnimation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+
+    startAnimation.addListener(_handleTick);
+    endAnimation.addListener(_handleTick);
+
+    final CurvedAnimation gapAnimation = new CurvedAnimation(
+      parent: controller,
+      curve: Curves.ease,
+      reverseCurve: Curves.ease
+    );
+
+    gapAnimation.addListener(_handleTick);
+
+    _animationTuples[gap.key] = new _AnimationTuple(
+      controller: controller,
+      startAnimation: startAnimation,
+      endAnimation: endAnimation,
+      gapAnimation: gapAnimation
+    );
+  }
+
+  @override
+  void initState() {
+    super.initState();
+    _children = new List<MergeableMaterialItem>.from(config.children);
+    _animationTuples = new Map<LocalKey, _AnimationTuple>();
+
+    for (int i = 0; i < _children.length; i += 1) {
+      if (_children[i] is MaterialGap) {
+        _initGap(_children[i]);
+        _animationTuples[_children[i].key].controller.value = 1.0; // Gaps must be wide when first built.
+      }
+    }
+    assert(_debugGapsAreValid(_children));
+  }
+
+  @override
+  void dispose() {
+    for (MergeableMaterialItem child in _children) {
+      if (child is MaterialGap) {
+        _animationTuples[child.key].controller.dispose();
+      }
+    }
+    super.dispose();
+  }
+
+  void _handleTick() {
+    setState(() {
+      // The animation's state is our build state, and it changed already.
+    });
+  }
+
+  bool _debugConsecutiveGaps(List<MergeableMaterialItem> children) {
```

Comment:
(otherwise it sounds like "debug" is the verb, and I'd expect it to launch a debugger or something)